src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -991,45 +999,86 @@
                                            [[package_name version provider] id]))
                                     (into {})))))
 
+(defn find-package-hashes [package-hashes]
+  (jdbc/call-with-query-rows [(format "SELECT id, %s as package_hash FROM packages WHERE hash = ANY(?)"
+                                      (sutils/sql-hash-as-str "hash"))
+                              (sutils/array-to-param "bytea" PGobject
+                                                     (map sutils/munge-hash-for-storage package-hashes))]
+                             (fn [rows]
+                               (reduce (fn [acc {:keys [id package_hash]}]
+                                         (assoc acc package_hash id))
+                                       {} rows))))
+
+(defn package-id-set-for-certname [certname-id]
+  (jdbc/call-with-query-rows ["SELECT package_id FROM certname_packages WHERE certname_id=?"
+                              certname-id]
+                             (fn [rows]
+                               (set (map :package_id rows)))))
+
+(defn insert-missing-packages [existing-hashes-map new-hashed-package-tuples]
+  (let [packages-to-create (remove (fn [hashed-package-tuple]
+                                     (get existing-hashes-map (facts/package-tuple-hash hashed-package-tuple)))
+                                   new-hashed-package-tuples)
+        results (jdbc/insert-multi! :packages
+                                    (map (fn [[package_name version provider package-hash]]
+                                           {:name package_name
+                                            :version version
+                                            :provider provider
+                                            :hash (sutils/munge-hash-for-storage package-hash)})
+                                         packages-to-create))]
+    (merge existing-hashes-map
+           (zipmap (map facts/package-tuple-hash packages-to-create)
+                   (map :id results)))))
+
 (s/defn update-packages
   "Compares `inventory` to the stored package inventory for
   `certname`. Differences will result in updates to the database"
   [certname
    inventory :- [facts/package-tuple]]
-  (let [{certname-id :id package_hash :package_hash} (find-certname-id-and-hash certname)
-        new-package-hash (shash/package-similarity-hash inventory)]
-
-    (when-not (= new-package-hash package_hash)
-      (let [old-packages (create-package-map certname-id)
-            [to-add to-remove _] (clojure.data/diff (set inventory) (set (keys old-packages)))
-            to-remove-ids (map old-packages to-remove)]
+  (let [{certname-id :id package-hash :package_hash} (find-certname-id-and-hash certname)
+        hashed-package-tuples (map shash/package-identity-hash inventory)
```

Comment:
I'm not seeing a guarantee that inventory is coming into this function in sorted order -- have we made that explicit somewhere?